### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by anchpop**
# Problem

We have no unified way of adding logs. See #64. Logs are often useful when debugging, but clutter up the console output otherwise, and it's annoying to have to remove and re-add them every time.

# Solutions

This PR adds the `Logs` OCaml library and sets it up for use within `sidecli` and `http_server`.  It also modifies all `sidecli` subcommands to take an option indicating that verbose logging should be enabled, and an option to control logging colors. (By default colors are on in sidecli, and vebosity is off.) Once I make `http_server` use `cmdliner` I will make the same change there.

### CODE REVIEW

File: bin/http_server.re [hunk 1]
```diff
@@ -15,6 +15,9 @@ open Protocol;
 open Node;
 open Networking;
 
+Logs.set_reporter(Logs_fmt.reporter());
+Logs.set_level(Logs.level_of_string("debug") |> Result.get_ok);
```
Comment (EduardoRFS) [inline bin/http_server.re:19]: I think there is a variant under `Logs` maybe `Logs.Debug`?

File: package.json [hunk 2]
```diff
@@ -56,4 +57,4 @@
     "@opam/opium": "github:EduardoRFS/opium:opium.opam#9fdbfbd0eedf238e1be0663ab95977be27d9a6e1",
     "@opam/rock": "github:EduardoRFS/opium:rock.opam#9fdbfbd0eedf238e1be0663ab95977be27d9a6e1"
   }
-}
+}
```
Comment (EduardoRFS) [inline package.json:60]: You need to fix your editor

**Review by EduardoRFS — COMMENTED** [review]
Not fully reviewed just some comments.

Overall I like the goal but I think there is a requirement to having known keys on logs and we also need to check how costly is the `Logs_lwt` vs `Logs`, my gut feeling tells me that `Logs_lwt` will not be worth on performance to us but I may be wrong.

**Comment by anchpop** [comment]
@EduardoRFS what do you mean by "having known keys on logs"?

And Logs_lwt should be basically a no-op if we disable logging, right? Or does the performance matter in other cases?

**Comment by EduardoRFS** [comment]
> And Logs_lwt should be basically a no-op if we disable logging, right?

I'm not sure if it is worth it the complexity of having a monad, I feel like the normal Logs is good enough.

> What do you mean by "having known keys on logs"?

A good way to ensure that we index things in a per module basis without affecting the performance that much.

File: bin/http_server.re [hunk 3]
```diff
@@ -15,6 +15,9 @@ open Protocol;
 open Node;
 open Networking;
 
+Logs.set_reporter(Logs_fmt.reporter());
+Logs.set_level(Some(Logs.Debug));
```
Comment (callistonianembrace) [inline bin/http_server.re:19]: I'd prefer if this is Error by default, perhaps Warning. By making it Debug, we treat all levels of the logs with the same priority

Comment (callistonianembrace) [reply bin/http_server.re:19]: Spoke to early: Line 19 shouldn't be necessary - Logs_cli should set the default level to warning already.

**Review by callistonianembrace — APPROVED** [review]
LGTM

Commented on setting of the default level as I didn't understand why it was done. I find value in the PR otherwise. LGTM

File: bin/http_server.re (hunk 3 above)
Comment (callistonianembrace) [reply bin/http_server.re:19]: Ignore me. http_server isn't using Logs_cli

**Comment by callistonianembrace** [comment]
This is a useful PR, @EduardoRFS. Does it need more work?

IMO, the title needs to reflect that the logs are not just added to sidecli.

**Comment by anchpop** [comment]
@callistonianembrace Edu mentioned that it would be nice if the logging included things like the file/line number the log was emitted at and the block height. The first requires I think a macro like the one @ulrikstrid wrote, and there's multiple ways to do the second but I thought the easiest would be to wait until we have an effect system

**Comment by EduardoRFS** [comment]
@anchpop please rebase this to begin.

File: bin/sidecli.re [hunk 4]
```diff
@@ -45,6 +59,8 @@ let exits =
 
 let lwt_ret = p => Term.(ret(const(Lwt_main.run) $ p));
 
+let const_log = p => Term.(const(() => p) $ setup_log);
```
Comment (d4hines) [inline bin/sidecli.re:62]: Can you explain the purpose of this function? It looks like it runs `setup_log`. Why do we have to run `setup_log` more than once?

Comment (anchpop) [reply bin/sidecli.re:62]: It should be only getting called once per sidecli invokation. If it's being called any more than that I've made a mistake

**Comment by anchpop** [comment]
@callistonianembrace 

> Commented on setting of the default level as I didn't understand why it was done. I find value in the PR otherwise. LGTM

That's a temporary measure until #113 is merged and the level can be set by cmdliner

**Review by d4hines — APPROVED** [review]
LGTM

**Review by EduardoRFS — CHANGES_REQUESTED** [review]
Two major problems, why `Logs_lwt`? What are we gaining on it? Seems like it should be removed.

At least we should have all logs message be prefixed by a key or something. Especially in a manner that can be easily filtered from the CLI, otherwise logging will become really annoying really fast.

Even if we don't have the line that is triggering the log this is a must have.

**Comment by anchpop** [comment]
@callistonianembrace I don't actually understand what Logs_lwt is even doing. Logging is normally a synchronous action, right? In what situation would Logs_lwt ever be preferred? I'm actually even more confused because [the docs say this](https://erratique.ch/software/logs/doc/Logs/index.html#sync) about the cooperative reporter example you linked:

> However a purely synchronous reporter like Reporter acting on channels does not play well with Lwt's cooperative runtime system. It is possible to reuse Reporter to define a cooperative reporter, see [this example](https://erratique.ch/software/logs/doc/Logs_lwt/index.html#report_ex). However while this reporter makes Logs_lwt's log functions synchronous, those of Logs behave asynchronously. For now it seems it that this is unfortunately the best we can do if we want to preserve the ability to use Logs with or without cooperative concurency.

And it seems very weird that Logs would behave asynchronously while Logs_lwt would behave synchronously (I would expect it to be the reverse). I can take out all the lwt stuff since @EduardoRFS says there's no benefit, but also I'd like to understand for what use case logs_lwt even exists at all

**Pull request closed without merging**